[nnc] Lazy lowerings registration; custom classes network params

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68151
* #68018
* #67711
* __->__ #67623

Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

```
  RegisterNNCLoweringsFunction quantized_conv2d_relu(
      {"quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)"},
      computeQuantizedConv2dRelu);
```
Ordering of lowerings registration and jit registration of custom classes `__torch__.torch.classes.quantized.Conv2dPackedParamsBase` depends on the order of linking compilation units and undefined.

So internal tests are failing with unknown jit class Conv2dPackedParamsBase.
To solve this prolem introducing lazy registration of lowerings wrapping them into functions that will be called once being called in initialization of `static` variable inside funtion.

IvanKobzarev committed:
[nnc] Lazy lowerings registration; custom classes network params

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32065076).

IvanKobzarev committed:
Update on "[nnc] Lazy lowerings registration; custom classes network params"


Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32065076).

IvanKobzarev committed:
Update on "[nnc] Lazy lowerings registration; custom classes network params"


Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

```
  RegisterNNCLoweringsFunction quantized_conv2d_relu(
      {"quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)"},
      computeQuantizedConv2dRelu);
```
Ordering of lowerings registration and jit registration of custom classes `__torch__.torch.classes.quantized.Conv2dPackedParamsBase` depends on the order of linking compilation units and undefined.

So internal tests are failing with unknown jit class Conv2dPackedParamsBase.
To solve this prolem introducing lazy registration of lowerings wrapping them into functions that will be called once being called in initialization of `static` variable inside funtion.

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32065076).

IvanKobzarev committed:
Update on "[nnc] Lazy lowerings registration; custom classes network params"


Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

```
  RegisterNNCLoweringsFunction quantized_conv2d_relu(
      {"quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)"},
      computeQuantizedConv2dRelu);
```
Ordering of lowerings registration and jit registration of custom classes `__torch__.torch.classes.quantized.Conv2dPackedParamsBase` depends on the order of linking compilation units and undefined.

So internal tests are failing with unknown jit class Conv2dPackedParamsBase.
To solve this prolem introducing lazy registration of lowerings wrapping them into functions that will be called once being called in initialization of `static` variable inside funtion.

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32065076).

IvanKobzarev committed:
Update on "[nnc] Lazy lowerings registration; custom classes network params"


Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

```
  RegisterNNCLoweringsFunction quantized_conv2d_relu(
      {"quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)"},
      computeQuantizedConv2dRelu);
```
Ordering of lowerings registration and jit registration of custom classes `__torch__.torch.classes.quantized.Conv2dPackedParamsBase` depends on the order of linking compilation units and undefined.

So internal tests are failing with unknown jit class Conv2dPackedParamsBase.
To solve this prolem introducing lazy registration of lowerings wrapping them into functions that will be called once being called in initialization of `static` variable inside funtion.

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32065076).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1103,11 +1103,14 @@ void TensorExprKernel::bindConstant(const torch::jit::Value* v) {
   if (torch::isCustomClass(val)) {
     auto name_hint = "const_" + sanitizeName(v->debugName());
-    auto dtype = Dtype(ScalarType::Float);
+    auto dtype = Dtype(ScalarType::Char);
```
Is this change needed?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1103,11 +1103,14 @@ void TensorExprKernel::bindConstant(const torch::jit::Value* v) {
     BufPtr buf = alloc<Buf>(name_hint, dims, dtype);
     auto dataPtr = val.toObjectRef().getSlot(0).toCapsule().get();
-    constants_.push_back({buf, dataPtr});
+    // NOLINTNEXTLINE
+    dims.push_back(alloc<LongImm>(sizeof(dataPtr)));
```
And this?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -88,7 +88,8 @@ std::vector<T> convertVecArgValue(const std::vector<ArgValue>& v) {
   struct ConstantDescr {
     BufPtr buf;
-    void* ptr;
+    void* ptr = nullptr;
+    Node* node = nullptr;
```
Could you please add a comment about these fields? I.e. that only one of them is expected to be non-null and what each of these situations means.

ZolotukhinM reviewed (approved):
Looks good, a couple of minor nits are below:

IvanKobzarev committed:
Update on "[nnc] Lazy lowerings registration; custom classes network params"


Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

```
  RegisterNNCLoweringsFunction quantized_conv2d_relu(
      {"quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)"},
      computeQuantizedConv2dRelu);
```
Ordering of lowerings registration and jit registration of custom classes `__torch__.torch.classes.quantized.Conv2dPackedParamsBase` depends on the order of linking compilation units and undefined.

So internal tests are failing with unknown jit class Conv2dPackedParamsBase.
To solve this prolem introducing lazy registration of lowerings wrapping them into functions that will be called once being called in initialization of `static` variable inside funtion.

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32065076).

IvanKobzarev committed:
Update on "[nnc] Lazy lowerings registration; custom classes network params"


Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

```
  RegisterNNCLoweringsFunction quantized_conv2d_relu(
      {"quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)"},
      computeQuantizedConv2dRelu);
```
Ordering of lowerings registration and jit registration of custom classes `__torch__.torch.classes.quantized.Conv2dPackedParamsBase` depends on the order of linking compilation units and undefined.

So internal tests are failing with unknown jit class Conv2dPackedParamsBase.
To solve this prolem introducing lazy registration of lowerings wrapping them into functions that will be called once being called in initialization of `static` variable inside funtion.

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32065076).

IvanKobzarev committed:
Update on "[nnc] Lazy lowerings registration; custom classes network params"


Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

```
  RegisterNNCLoweringsFunction quantized_conv2d_relu(
      {"quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)"},
      computeQuantizedConv2dRelu);
```
Ordering of lowerings registration and jit registration of custom classes `__torch__.torch.classes.quantized.Conv2dPackedParamsBase` depends on the order of linking compilation units and undefined.

So internal tests are failing with unknown jit class Conv2dPackedParamsBase.
To solve this prolem introducing lazy registration of lowerings wrapping them into functions that will be called once being called in initialization of `static` variable inside funtion.

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32065076).

IvanKobzarev committed:
Update on "[nnc] Lazy lowerings registration; custom classes network params"


Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

```
  RegisterNNCLoweringsFunction quantized_conv2d_relu(
      {"quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)"},
      computeQuantizedConv2dRelu);
```
Ordering of lowerings registration and jit registration of custom classes `__torch__.torch.classes.quantized.Conv2dPackedParamsBase` depends on the order of linking compilation units and undefined.

So internal tests are failing with unknown jit class Conv2dPackedParamsBase.
To solve this prolem introducing lazy registration of lowerings wrapping them into functions that will be called once being called in initialization of `static` variable inside funtion.

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32065076).

IvanKobzarev committed:
Update on "[nnc] Lazy lowerings registration; custom classes network params"


Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

```
  RegisterNNCLoweringsFunction quantized_conv2d_relu(
      {"quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)"},
      computeQuantizedConv2dRelu);
```
Ordering of lowerings registration and jit registration of custom classes `__torch__.torch.classes.quantized.Conv2dPackedParamsBase` depends on the order of linking compilation units and undefined.

So internal tests are failing with unknown jit class Conv2dPackedParamsBase.
To solve this prolem introducing lazy registration of lowerings wrapping them into functions that will be called once being called in initialization of `static` variable inside funtion.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[nnc] Lazy lowerings registration; custom classes network params"


Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

```
  RegisterNNCLoweringsFunction quantized_conv2d_relu(
      {"quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)"},
      computeQuantizedConv2dRelu);
```
Ordering of lowerings registration and jit registration of custom classes `__torch__.torch.classes.quantized.Conv2dPackedParamsBase` depends on the order of linking compilation units and undefined.

So internal tests are failing with unknown jit class Conv2dPackedParamsBase.
To solve this prolem introducing lazy registration of lowerings wrapping them into functions that will be called once being called in initialization of `static` variable inside funtion.

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32065076).

IvanKobzarev committed:
Update on "[nnc] Lazy lowerings registration; custom classes network params"


Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

```
  RegisterNNCLoweringsFunction quantized_conv2d_relu(
      {"quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)"},
      computeQuantizedConv2dRelu);
```
Ordering of lowerings registration and jit registration of custom classes `__torch__.torch.classes.quantized.Conv2dPackedParamsBase` depends on the order of linking compilation units and undefined.

So internal tests are failing with unknown jit class Conv2dPackedParamsBase.
To solve this prolem introducing lazy registration of lowerings wrapping them into functions that will be called once being called in initialization of `static` variable inside funtion.

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32065076).

IvanKobzarev committed:
Update on "[nnc] Lazy lowerings registration; custom classes network params"


Differential Revision: [D32065076](https://our.internmc.facebook.com/intern/diff/D32065076)

```
  RegisterNNCLoweringsFunction quantized_conv2d_relu(
      {"quantized::conv2d_relu.new(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> (Tensor)"},
      computeQuantizedConv2dRelu);
```
Ordering of lowerings registration and jit registration of custom classes `__torch__.torch.classes.quantized.Conv2dPackedParamsBase` depends on the order of linking compilation units and undefined.

So internal tests are failing with unknown jit class Conv2dPackedParamsBase.
To solve this prolem introducing lazy registration of lowerings wrapping them into functions that will be called once being called in initialization of `static` variable inside funtion.

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32065076).
